wasm/src/thread_pool/mod.rs

```diff
@@ -0,0 +1,180 @@
+
+    let (sender, receiver) = channel();
+
+    let receiver = Box::leak(Box::new(receiver));
+
+    let workers = try_join_all((0..num_threads).map(|_| {
+        JsFuture::from(spawn_worker(&url, &module, &memory, receiver))
+    })).await?;
+
+    // Needed to work around a Firefox bug where Workers get garbage collected too early
+    // https://bugzilla.mozilla.org/show_bug.cgi?id=1592227
+    std::mem::forget(workers);
+
+    Ok(sender)
+}
+
+async fn spawn_local_thread_pool(url: web_sys::Url, num_threads: usize) -> Result<rayon::ThreadPool, JsValue> {
+    let pool;
+
+    if num_threads == 1 {
+        pool = rayon::ThreadPoolBuilder::new()
+            .num_threads(1)
+            .use_current_thread()
+            .build()
+            .unwrap_throw();
+
+    } else {
+        let mut sender = spawn_workers(url, num_threads).await?;
+
+        pool = rayon::ThreadPoolBuilder::new()
+            .num_threads(num_threads)
+            .spawn_handler(move |thread| {
+                sender.send(thread).unwrap_throw();
+                Ok(())
+            })
+            .build()
+            .unwrap_throw();
+    }
+
+    Ok(pool)
+}
+
+async fn spawn_global_thread_pool(url: web_sys::Url, num_threads: usize) -> Result<(), JsValue> {
+    if num_threads == 1 {
+        rayon::ThreadPoolBuilder::new()
+            .num_threads(1)
+            .use_current_thread()
+            .build_global()
+            .unwrap_throw();
+
+    } else {
+        let mut sender = spawn_workers(url, num_threads).await?;
+
+        rayon::ThreadPoolBuilder::new()
+            .num_threads(num_threads)
+            .spawn_handler(move |thread| {
+                sender.send(thread).unwrap_throw();
+                Ok(())
+            })
+            .build_global()
+            .unwrap_throw();
+    }
+
+    Ok(())
+}
+
+
+pub struct ThreadPool {
+    url: Option<web_sys::Url>,
+    num_threads: Option<usize>,
+}
+
+impl ThreadPool {
+    pub fn builder() -> Self {
+        Self {
+            url: None,
+            num_threads: None,
+        }
+    }
+
+    pub fn url(mut self, url: web_sys::Url) -> Self {
+        self.url = Some(url);
+        self
+    }
+
+    pub fn num_threads(mut self, num_threads: usize) -> Self {
+        self.num_threads = Some(num_threads);
+        self
+    }
+
+    fn defaults(self) -> (web_sys::Url, usize) {
+        (
+            self.url.expect("Missing url for ThreadPool"),
+
+            self.num_threads.unwrap_or_else(|| {
+                let window: web_sys::Window = js_sys::global().unchecked_into();
+                window.navigator().hardware_concurrency() as usize
+            }),
+        )
+    }
+
+    // TODO this should cleanup the receiver when the ThreadPool is dropped
+    pub fn build_local(self) -> impl Future<Output = Result<rayon::ThreadPool, JsValue>> {
+        let (url, num_threads) = self.defaults();
+        spawn_local_thread_pool(url, num_threads)
+    }
```

Comment:
I like this approach. When this is cleaned up I assume the workers are also dropped?

Reply:
Currently no, but I have a TODO to fix that.